crates/sparrow-backend/src/pipeline_schedule.rs

```diff
@@ -0,0 +1,163 @@
+use sparrow_core::debug_println;
+use sparrow_physical::{Pipeline, Step, StepKind};
+
+const DEBUG_SCHEDULING: bool = false;
+
+/// Determine the pipeline each step should be part of.
+///
+/// This is computed in a bottom-up manner:
+///
+/// 1. Each leaf (input) is the start of a separate pipeline.
+/// 2. Certain step kinds are "pipeline breaking", which means
+///    they start a new pipeline. For instance, `with_key`.
+/// 3. Any other operation with a single input is part of the
+///    the same pipeline as the input.
+/// 4. Any other operation is a separate pipeline.
+pub fn pipeline_schedule(steps: &[Step]) -> Vec<Pipeline> {
+    // Compute how many times each step is referenced.
+    // Any step referenced multiple times must end a pipeline.
```

Comment:
I don't quite understand this rule. Wouldn't the `sum` step be referenced multiple times, but still possible to be part of the same pipeline? 
```
let sum = sum(Foo)
in { f1: sum, f2: sum }
```

Reply:
The `sum` is an aggregation -- it's an expression within one step. In the above, we'd actually only have one step performing a aggregations.

```
Step {
  kind: StepKind::Aggregation {
    aggregates: vec![
      0="inputs[0]",
      1="sum(0)",
      2="last(1)",
      3="first(1)",
  ],
  inputs: vec![scan],
  schema: ...
}
```